orx-palette/src/main/kotlin/PaletteStudio.kt

```diff
@@ -0,0 +1,219 @@
+            logger.error(ex) { "Error: Could not load palettes" }
+            logger.info { "Only JSON files with Array<Array<String>> structure can be loaded using this method" }
+        }
+    }
+
+    private fun loadFromURL(url: URL): Unit = load(url.readText())
+
+    private fun createPalette(colors: MutableList<ColorRGBa>) : Palette {
+        when(sortBy) {
+            SortBy.DARKEST -> {
+                val darkest = Comparator<ColorRGBa> { c1: ColorRGBa, c2: ColorRGBa -> (getLuminance(c1) - getLuminance(c2)).toInt() }
+                colors.sortWith(darkest)
+            }
+            SortBy.BRIGHTEST -> {
+                val brightest = Comparator<ColorRGBa> { c1: ColorRGBa, c2: ColorRGBa -> (getLuminance(c2) - getLuminance(c1)).toInt() }
+                colors.sortWith(brightest)
+            }
+            SortBy.NO_SORTING -> {}
+        }
+
+        val background = colors.first()
+
+        val foreground = colors
+                .takeLast(colors.size - 1)
+                .map { getContrast(background, it) to it }
+                .maxBy { it.first }!!
+                .second
+
+        var constraint =  colors.size
+        var constraint2 = colors.size
+
+        if (colorCountConstraint > 0 && colorCountConstraint < colors.size) {
+            constraint = colorCountConstraint
+            constraint2 = colorCountConstraint + 1
+        }
+
+        val colors1 = colors.slice(0 until constraint)
+        val colors2 = colors.slice(1 until constraint2)
+
+        return Palette(background, foreground, colors1, colors2)
+    }
+
+    fun loadExternal(filePath: String) {
+        palettes = mutableListOf()
+
+        load(File(filePath).readText())
+
+        val choice = Random.pick(palettes)
+
+        palette = createPalette(choice)
+        paletteIndex = palettes.indexOf(choice)
+    }
+
+    fun select(index: Int = 0) {
+        paletteIndex = max(min(index, palettes.size - 1), 0)
+        palette = createPalette(palettes[paletteIndex])
+    }
+
+    fun getIndex(): Int {
+        return paletteIndex
+    }
+
+    fun randomize() {
+        palette = createPalette(Random.pick(palette.colors, count = palette.colors.size))
+    }
+
+    fun randomPalette() {
+        val comparison = palette.colors.toMutableList()
+        val colors= Random.pick(palettes, comparison) as MutableList<ColorRGBa>
+
+        paletteIndex = palettes.indexOf(colors)
+        palette = createPalette(colors)
+    }
+
+    fun changeCollection(newCollection: Collections) {
+        loadCollection(newCollection)
+    }
+
+    private fun registerKeybindings(keyboard: Keyboard) {
+        keyboard.keyDown.listen {
+            if (it.name == "l") {
+                randomPalette()
+            }
+            if (it.name == "k") {
+                randomize()
+            }
+        }
+    }
+
+    /*
+     * EXTENSION
+     */
+    override var enabled: Boolean = true
+
+    override fun setup(program: Program) {
```

Comment:
When I think about this, would it be possible to make these key bindings optional? I imagine that I might want to use this extension and I might have already some assignments for these keys.

Reply:
They are optional, they just get added if you use `extend(paletteStudio)`.